tests/arm7tdmi/arm7tdmi.cpp

```diff
@@ -0,0 +1,522 @@
+    if(mode & Byte) return 1;
+    if(mode & Half) return 2;
+    if(mode & Word) return 4;
+    print("unrecognized mode ", hex(mode, 3L), "\n");
+    abort();
+    return 4;
+  }
+
+  auto matchTransaction(u32 mode, n32 address, n32 word = 0) -> maybe<u32> {
+    maybe<u32> result;
+    u32 size = memSize(mode);
+    u32 data = word & (~0u >> (32 - size * 8));
+    if(tindex < test->transactions.size()) {
+      auto& t = test->transactions[tindex++];
+      if((mode & Prefetch) && t.kind == 0 && t.size == size && t.addr == address) {
+        result = t.data;
+      }
+      if((mode & Load) && t.kind == 1 && t.size == size && t.addr == address) {
+        result = t.data;
+      }
+      if((mode & Store) && t.kind == 2 && t.size == size && t.addr == address) {
+        if(t.data == data) {
+          result = t.data;
+        } else {
+          error("write: data ", hex(data, 2 * size), " != ", hex(t.data, 2 * size), ")\n");
+        }
+      }
+      if(result) {
+        //don't check sequential unless we got a result, otherwise it's just noise
+        if(!(mode & Sequential) != !(t.access & Access::Sequential)) {
+          error("write: nonsequential ", !(mode & Sequential), " != ", !(t.access & Access::Sequential), "\n");
+          result = nothing;
+        }
+      }
+    }
+    return result;
+  }
+
+  auto run(const TestCase& test, bool logErrors) -> TestResult;
+
+  template<typename... P>
+  auto error(P&&... p) -> void {
+    errors.append(string{std::forward<P>(p)...});
+  }
+} cpu;
+
+auto CPU::run(const TestCase& test, bool logErrors) -> TestResult {
+  const auto& is = test.initial;
+  const auto& fs = test.final;
+
+  power(test);
+
+  processor.r0 = is.R[0];
+  processor.r1 = is.R[1];
+  processor.r2 = is.R[2];
+  processor.r3 = is.R[3];
+  processor.r4 = is.R[4];
+  processor.r5 = is.R[5];
+  processor.r6 = is.R[6];
+  processor.r7 = is.R[7];
+  processor.r8 = is.R[8];
+  processor.r9 = is.R[9];
+  processor.r10 = is.R[10];
+  processor.r11 = is.R[11];
+  processor.r12 = is.R[12];
+  processor.r13 = is.R[13];
+  processor.r14 = is.R[14];
+  processor.r15 = is.R[15];
+  processor.fiq.r8 = is.R_fiq[0];
+  processor.fiq.r9 = is.R_fiq[1];
+  processor.fiq.r10 = is.R_fiq[2];
+  processor.fiq.r11 = is.R_fiq[3];
+  processor.fiq.r12 = is.R_fiq[4];
+  processor.fiq.r13 = is.R_fiq[5];
+  processor.fiq.r14 = is.R_fiq[6];
+  processor.svc.r13 = is.R_svc[0];
+  processor.svc.r14 = is.R_svc[1];
+  processor.abt.r13 = is.R_abt[0];
+  processor.abt.r14 = is.R_abt[1];
+  processor.irq.r13 = is.R_irq[0];
+  processor.irq.r14 = is.R_irq[1];
+  processor.und.r13 = is.R_und[0];
+  processor.und.r14 = is.R_und[1];
+  processor.cpsr = is.CPSR;
+  processor.fiq.spsr = is.SPSR[0];
+  processor.svc.spsr = is.SPSR[1];
+  processor.abt.spsr = is.SPSR[2];
+  processor.irq.spsr = is.SPSR[3];
+  processor.und.spsr = is.SPSR[4];
+
+  const bool thumb = processor.cpsr.t;
+  const u32 length = thumb ? 2 : 4;
+
+  //tests/v1/arm_mrs.json
```

Comment:
Just wanted to note the obvious: it would be nicer if these exceptions were data-driven instead of hardcoded. However, the end goal is to improve the tests (and ares in the process) and get rid of them entirely.